conans/client/graph/graph_binaries.py

```diff
@@ -156,13 +157,18 @@ def _evaluate_node(self, node, build_mode, update, remotes):
 
         # If it has lock
         locked = node.graph_lock_node
-        if locked and locked.pref.id == node.package_id:
-            pref = locked.pref  # Keep the locked with PREV
+        if locked and locked.package_id:  # if package_id = None, nothing to lock here
+            # First we update the package_id, just in case there are differences or something
+            locked.package_id = node.package_id  # necessary for PACKAGE_ID_UNKNOWN
+            pref = PackageReference(locked.ref, locked.package_id, locked.prev)  # Keep locked PREV
             self._process_node(node, pref, build_mode, update, remotes)
             if node.binary == BINARY_MISSING and build_mode.allowed(node.conanfile):
                 node.binary = BINARY_BUILD
+            if node.binary == BINARY_BUILD and locked.prev:
+                raise ConanException("Trying to build '%s', but it is locked" % repr(node.ref))
```

Comment:
Maybe we could give more detailed information here, something like:

_"Cant build 'liba/1.0' because it is already locked in the input lockfile to prev='7e26fb70378aef57837fc0957376e91f' of 'liba/1.0#5f052cb9b7bc8026786b7a88aaf9ab88'"_